API for querying available memory + cgroup1 fixes

maleadt opened the pull request:
This PR mimicks https://github.com/libuv/libuv/pull/3744, accessing `memory.usage_in_bytes`/`memory.current` (cgroupv1/cgroupv2) and using it to implement a constraint-aware version of `uv_get_free_memory`.

Implements https://github.com/libuv/libuv/pull/3744#issuecomment-1249464780

bnoordhuis commented on src/unix/linux-core.c:
```diff
@@ -1228,6 +1228,57 @@ uint64_t uv_get_constrained_memory(void) {
+  if (memcmp(buf, "0::/", 4))
+    return uv__get_available_memory_fallback();
+
+  total = uv_get_total_memory();
+  if (constrained > total)
+    return uv_get_free_memory();
```
I was going to suggest moving this code up a few lines so you don't read & parse /proc/self/cgroup for nothing but I realize that's not 100% equivalent because these lines won't be reached unless we're on a cgroupsv2 system.

That's fairly subtle assuming it's intentional so please at least add a comment explaining the what and why.

bnoordhuis commented on src/unix/linux-core.c:
```diff
@@ -1228,6 +1228,57 @@ uint64_t uv_get_constrained_memory(void) {
+  snprintf(filename, sizeof(filename), "/sys/fs/cgroup/%s/memory.current", p);
+  current = uv__read_uint64(filename);
+
+  return constrained - current;
```
Should probably have the same underflow guard as the fallback code.

bnoordhuis commented on test/test-get-memory.c:
```diff
@@ -26,11 +26,13 @@ TEST_IMPL(get_memory) {
+  uint64_t available_mem = uv_get_available_memory();
 
-  printf("free_mem=%llu, total_mem=%llu, constrained_mem=%llu\n",
+  printf("free_mem=%llu, total_mem=%llu, constrained_mem=%llu, available_memo=%llu\n",
```
```suggestion
  printf("free_mem=%llu, total_mem=%llu, constrained_mem=%llu, available_mem=%llu\n",
```
And please wrap at 80 columns.

bnoordhuis commented on test/test-get-memory.c:
```diff
@@ -40,5 +42,6 @@ TEST_IMPL(get_memory) {
 #else
   ASSERT(total_mem > free_mem);
 #endif
+  ASSERT(available_mem <= free_mem);
```
```suggestion
  ASSERT_LE(available_mem, free_mem);
```
New code should use the new-style macros. I really should move everything ever to the new macros sometime but blergh, so tedious.

bnoordhuis reviewed (commented):
Thanks for the PR.

bnoordhuis commented:
by the way, you probably want to rebase if you're having CI trouble. Also, linux-core.c is about to be renamed to linux.c and merged with some other files. There's a pretty good chance git will handle it just fine though.

maleadt commented:
Thanks for the review. I've addressed your comments and rebased the PR. To get rid of some of the subtleness and make the code easier to read, I've refactored the get_constrained and get_available methods significantly.

Now that this PR contains quite a bit of a refactor anyway, should I fold the other pending changes (https://github.com/libuv/libuv/pull/3753 for detecting the cgroupv1 max limit, and https://github.com/JuliaLang/libuv/pull/27 with a fix for cgroupv1 limits) into here?

bnoordhuis commented on src/unix/linux.c:
```diff
@@ -1327,47 +1327,112 @@ static uint64_t uv__read_uint64(const char* filename) {
+  */
+  *high = uv__read_uint64("/sys/fs/cgroup/memory/memory.soft_limit_in_bytes");
+  *max = uv__read_uint64("/sys/fs/cgroup/memory/memory.limit_in_bytes");
+
+  return;
```
```suggestion
```

bnoordhuis commented on src/unix/linux.c:
```diff
@@ -1327,47 +1327,112 @@ static uint64_t uv__read_uint64(const char* filename) {
+    return 0;
+
+  return high < max ? high : max;
+
```
```suggestion
```

bnoordhuis commented on src/unix/linux.c:
```diff
@@ -1327,47 +1327,112 @@ static uint64_t uv__read_uint64(const char* filename) {
+    return uv_get_free_memory();
+
+  /* In the case of cgroupv2, we'll only have a single entry. */
+  if (0 == memcmp(buf, "0::/", 4)) {
+    current = uv__get_cgroup2_current_memory(buf);
+  } else {
+    current = uv__get_cgroup1_current_memory(buf);
+  }
```
```suggestion
  if (0 == memcmp(buf, "0::/", 4))
    current = uv__get_cgroup2_current_memory(buf);
  else
    current = uv__get_cgroup1_current_memory(buf);
```

bnoordhuis commented on src/unix/linux.c:
```diff
@@ -1327,47 +1327,112 @@ static uint64_t uv__read_uint64(const char* filename) {
+
+  snprintf(filename, sizeof(filename), "/sys/fs/cgroup/%s/memory.high", p);
+  *high = uv__read_uint64(filename);
+
+  return;
```
```suggestion
```

bnoordhuis reviewed (approved):
Style nits but otherwise LGTM, thanks!

Maybe one code improvement you could consider: changing the functions that take `char* buf` to `char (*buf)[1024]` - then it's always clear what size the buffer is.

maleadt commented:
> Maybe one code improvement you could consider: changing the functions that take `char* buf` to `char (*buf)[1024]` - then it's always clear what size the buffer is.

Doing so then requires a bunch of casts to avoid warnings on e.g. `strchr` (`expected ‘const char *’ but argument is of type ‘char (*)[1024]’`), so I left that out for now, but I can put it back if you want.

I'll rebase my other cgroup-related PRs on top of this one.

maleadt commented:
I also pushed a commit that handles the case where with cgroup1 the memory limits are applied in a subgroup. This was reported by @simonbyrne, who noticed that with SLURM the memory limits are set in the memory subgroup that's listed in `/proc/self/cgroup`:

```
├ cat /proc/self/cgroup # cgroups of current process
11:blkio:/
10:memory:/slurm/uid_5184/job_28925944/step_0
9:freezer:/slurm/uid_5184/job_28925944/step_0
8:pids:/
7:cpuset:/slurm/uid_5184/job_28925944/step_0
6:net_prio,net_cls:/
5:perf_event:/
4:hugetlb:/
3:devices:/slurm/uid_5184/job_28925944/step_0/task_0
2:cpuacct,cpu:/
1:name=systemd:/system.slice/slurmd.service

├ cat /sys/fs/cgroup/memory/slurm/uid_5184/job_28925944/step_0/memory.limit_in_bytes # applied limit
2147483648

├ cat /sys/fs/cgroup/memory/memory.limit_in_bytes # what we currently read (unconstrained)
9223372036854771712
```

I'm not entirely sure the way I'm doing things is correct though, because I cannot replicate the above. In fact, with Docker on a cgroup1 system the memory limit seems applied to the top level of the memory controller hierarchy, and the subgroup isn't even accessible:

```
$ docker run --rm -it --memory="3g" archlinux

# cat /proc/self/cgroup
12:rdma:/
11:hugetlb:/docker/86ca392018812da81965d84e0331219321a2a4977d325324f6fa939de3200d97
10:blkio:/docker/86ca392018812da81965d84e0331219321a2a4977d325324f6fa939de3200d97
9:freezer:/docker/86ca392018812da81965d84e0331219321a2a4977d325324f6fa939de3200d97
8:devices:/docker/86ca392018812da81965d84e0331219321a2a4977d325324f6fa939de3200d97
7:cpuset:/docker/86ca392018812da81965d84e0331219321a2a4977d325324f6fa939de3200d97
6:memory:/docker/86ca392018812da81965d84e0331219321a2a4977d325324f6fa939de3200d97
5:cpu,cpuacct:/docker/86ca392018812da81965d84e0331219321a2a4977d325324f6fa939de3200d97
4:pids:/docker/86ca392018812da81965d84e0331219321a2a4977d325324f6fa939de3200d97
3:net_cls,net_prio:/docker/86ca392018812da81965d84e0331219321a2a4977d325324f6fa939de3200d97
2:perf_event:/docker/86ca392018812da81965d84e0331219321a2a4977d325324f6fa939de3200d97
1:name=systemd:/docker/86ca392018812da81965d84e0331219321a2a4977d325324f6fa939de3200d97
0::/system.slice/containerd.service

# ls /sys/fs/cgroup/memory/docker
ls: cannot access '/sys/fs/cgroup/memory/docker': No such file or directory

# cat /sys/fs/cgroup/memory/memory.limit_in_bytes
3221225472
```

Can put this in a separate PR if you prefer.

maleadt commented:
I also relaxed the `ASSERT_LE` test, now checking that the available memory is <= the total memory (and not the free memory), because such a check is fragile (and fails when memory is released between the test suite's calls to get_free and get_available):

```
not ok 1 - get_memory
# exit code 6
# Output from process `get_memory`:
# free_mem=34517368832, total_mem=66810916864, constrained_mem=18446744073709551615, available_mem=34517864448
# Assertion failed in /home/tim/Julia/src/libuv/test/test-get-memory.c on line 46: `available_mem <= free_mem` (34517864448 <= 34517368832)
```

simonbyrne commented:
I can confirm this works correctly on our Slurm cluster:
```
┌─[25]──[Wed Oct 26]──[10:10:39]────────────────────────────────────────
│ spjbyrne@login1:~/src/libuvx
├ julia -E 'ccall((:uv_get_constrained_memory,".libs/libuv.so"), UInt64, ())'
0x7ffffffffffff000
┌─[26]──[Wed Oct 26]──[10:10:48]────────────────────────────────────────
│ spjbyrne@login1:~/src/libuvx
├ srun -t 00:10:00 --mem 2G julia -E 'ccall((:uv_get_constrained_memory,".libs/libuv.so"), UInt64, ())'
0x0000000080000000
```

simonbyrne commented:
@maleadt in your docker container, what does:
```
cgget --variable memory.limit_in_bytes /docker/86ca392018812da81965d84e0331219321a2a4977d325324f6fa939de3200d97
```
give?

maleadt commented:
```
# grep memory /proc/self/cgroup
6:memory:/docker/e53d41e9e88db7741099fd1a6d93e7ab6299ea8ae7557ec4892c12217d4e5a42

# cgget --variable memory.limit_in_bytes /docker/e53d41e9e88db7741099fd1a6d93e7ab6299ea8ae7557ec4892c12217d4e5a42
variable file read failed No such file or directory
```

It tries to open the same file we've been considering, and nothing more:

```
openat(AT_FDCWD, "/sys/fs/cgroup/memory/docker/e53d41e9e88db7741099fd1a6d93e7ab6299ea8ae7557ec4892c12217d4e5a42//memory.limit_in_bytes", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
```

maleadt commented:
@bnoordhuis Do you have the time to give this another round of review?

bnoordhuis commented on src/unix/linux.c:
```diff
@@ -1341,47 +1341,139 @@ static uint64_t uv__read_uint64(const char* filename) {
+
+  if (p != NULL) {
+    p = p + 8;
+    p_end = strchr(p, '\n');
+    if (p_end != NULL)
+      *p_end = '\0';
+
+    /* If the memory controller is mounted, check limits there. */
+    snprintf(filename, sizeof(filename), "/sys/fs/cgroup/memory/%s/", p);
```
I realize this is an existing idiom but it'd be neater to not mutate in place.
```suggestion
    /* If the memory controller is mounted, check limits there. */
    n = (int) strcspn(p, "\n");
    snprintf(filename, sizeof(filename), "/sys/fs/cgroup/memory/%.*s/", n, p);
```

bnoordhuis commented on src/unix/linux.c:
```diff
@@ -1341,47 +1341,139 @@ static uint64_t uv__read_uint64(const char* filename) {
+
+    /* If the memory controller is mounted, check limits there. */
+    snprintf(filename, sizeof(filename), "/sys/fs/cgroup/memory/%s/", p);
+    if (0 == access(filename, F_OK)) {
```
This is a (admittedly mostly hypothetical) TOCTOU issue because the directory can get unmounted after access() but before open(). I'd be happier if it simply tried to read and handle the contingency if the files don't exist.

bnoordhuis commented on src/unix/linux.c:
```diff
@@ -1341,47 +1341,139 @@ static uint64_t uv__read_uint64(const char* filename) {
+  if (p != NULL)
+    *p = '\0';
+
+  p = buf + 4;
```
There's a bug here. I think you think you're skipping over `0::/` but `p` is either null or pointing to `\n` at this point.

maleadt replied:
I think this works as intended, as it does `buf + 4`, not `p + 4`. But I'll change it to the non-mutating version as you suggested above.

bnoordhuis commented on src/unix/linux.c:
```diff
@@ -1341,47 +1341,139 @@ static uint64_t uv__read_uint64(const char* filename) {
-  char* p;
+
+  /* In the case of cgroupv2, we'll only have a single entry. */
+  if (0 == memcmp(buf, "0::/", 4))
```
Potentially reads into uninitialized memory. Use strncmp.

bnoordhuis commented on src/unix/linux.c:
```diff
@@ -1341,47 +1341,139 @@ static uint64_t uv__read_uint64(const char* filename) {
 
-  if (memcmp(buf, "0::/", 4))
-    return uv__get_constrained_memory_fallback();
+static uint64_t uv__get_cgroup2_current_memory(char* buf) {
```
All these `char* buf` parameters make me kind of twitchy. As a callee, you don't really know how big the buffer is, or if it's nul-terminated.

C89 idioms for addressing that are `char* buf, size_t n` (specify length) or `char (*buf)[1024]` (specify array of exactly 1024 elements).

C99 lets you write `char buf[static 1024]`, meaning an array of _at least_ 1024 elements. Libuv is effectively C11 now so using that idiom would be okay.

maleadt replied:
Didn't know about the C99 syntax, that one works perfectly (i.e. without requiring lots of casts, as seemed required with `char (*buf)[1024]`).

maleadt commented on src/unix/linux.c:
```diff
@@ -1341,47 +1341,139 @@ static uint64_t uv__read_uint64(const char* filename) {
+    p = strchr(p, '\n');
+    if (p != NULL)
+      p = strchr(p, ':');
+  }
```
I don't think so, as `:memory:` can theoretically be part of the controller path:

```
$ sudo mkdir /sys/fs/cgroup/freezer/:memory:
# echo $$ > /sys/fs/cgroup/freezer/:memory:/cgroup.procs
$ cat /proc/self/cgroup | grep freezer
9:freezer:/:memory:
```

maleadt commented:
Rebased and addressed review comments.

bnoordhuis commented on src/unix/linux.c:
```diff
@@ -1343,47 +1343,163 @@ static uint64_t uv__read_uint64(const char* filename) {
 
+  if (p != NULL) {
+    /* Find out where the controller is mounted. */
+    p = p + 9;  /* skip :memory:/ */
```
```suggestion
    p = p + sizeof(":memory:");
```
Just a suggestion but it's self-documenting.

maleadt replied:
Ah yeah, that's neat. I used `strlen` instead so that we don't have to do `-1` or remove the trailing `/` (I verified that this is optimized away with my host gcc compiler).

bnoordhuis commented on src/unix/linux.c:
```diff
@@ -1343,47 +1343,163 @@ static uint64_t uv__read_uint64(const char* filename) {
+  int n;
+
+  /* Find out where the controller is mounted. */
+  p = buf + 4;  /* skip 0::/ */
```
```suggestion
  p = buf + sizeof("0::/") - 1;
```
Ditto.

bnoordhuis commented on src/unix/linux.c:
```diff
@@ -1343,47 +1343,163 @@ static uint64_t uv__read_uint64(const char* filename) {
+  if (p != NULL) {
+    /* Find out where the controller is mounted. */
+    p = p + 9;  /* skip :memory:/ */
+    n = (int) strcspn(p, "\n");
```
Can you factor this out into something like this:
```c
static char* uv__cgroup1_find_mount(char buf[static 1024], int* n) {
  // ...
}
```

bnoordhuis reviewed (commented):
Almost LGTM :-)

bnoordhuis reviewed (approved):
Thanks, LGTM!
